cpp/src/dataframe/types.hpp

```diff
@@ -38,74 +39,161 @@
 /* ----------------------------------------------------------------------------*/
 namespace cudf
 {
-  struct category
+
+namespace detail
+{
+/**
+     * @brief Base wrapper structure to emulate "strong typedefs" for gdf_dtype values 
+     * that do not correspond to fundamental types.
+     * 
+     * Implements operators that allow the wrapper to be used as if it were a fundamental
+     * type.
+     * 
+     * @tparam T  The type of the wrapped value
+     * @tparam type_id  The wrapped gdf_dtype
+     */
+template <typename T, gdf_dtype type_id>
+struct wrapper
+{
+  static constexpr gdf_dtype element_type_id{type_id}; ///< The wrapped gdf_dtype
+  using value_type = T;                                ///< The underlying fundamental type of the wrapper
+  value_type value;                                    ///< The wrapped value
+
+  CUDA_HOST_DEVICE_CALLABLE
+  explicit wrapper(T v): value{v} {}
+
+  CUDA_HOST_DEVICE_CALLABLE
+  wrapper(wrapper const& w): value(w.value) {}
+
+  CUDA_HOST_DEVICE_CALLABLE
+  wrapper(): value{value_type(0)}{}
+
+  CUDA_HOST_DEVICE_CALLABLE
+  explicit operator value_type() const { return this->value; }
+
+  CUDA_HOST_DEVICE_CALLABLE
+  wrapper& operator=(wrapper const &w)
```

Comment:
It's better to place these operators outside the class (but in the same namespace) unless they need access to protected members and methods. Same comment for all such operators - they can just be freestanding functions.

Reply:
Done.